driver: imx: Add PDM MICFIL driver

dbaluta opened the pull request:
The Pulse Density Modulated Microphone Interface (MICFIL) is a popular way to deliver audio from microphones to the processor in several applications, such as mobile telephones. However, current digital-audio systems use multibit audio signal (also known as multibit PCM) to represent the signal. This block implements the required digital interface to provide a 24-bits audio signal from a PDM microphone bitstream in a configurable output sampling rate.

This patch adds initial support for PDM MICFIL IP found on i.MX8MP board.

Follow PRs will add:
- topology for PDM MICFIL on imx8mp
- support for multi fifo script

lgirdwood reviewed (approved):
@dbaluta I guess this also needs to land in Zephyr at some point ?

dbaluta commented:
> @dbaluta I guess this also needs to land in Zephyr at some point ?

Yes. Sometimes in Q1 2024.

lgirdwood commented:
@iuliana-prodan good for you ?

lgirdwood commented:
@LaurentiuM1234 good for you ?

kv2019i commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,330 @@
+
+	dai_update_bits(dai, REG_MICFIL_CTRL1, MICFIL_CTRL1_DISEL,
+			MICFIL_CTRL1_DISEL_BITS(MICFIL_CTRL1_DISEL_DISABLE));
+	return 0;
```
Empty line before return missing.

kv2019i reviewed (approved):
Looks good!

RanderWang commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,331 @@
+	case COMP_TRIGGER_PRE_RELEASE:
+		break;
+	default:
+		dai_err(dai, "MICFIL: invalid trigger cmd %d", cmd);
```
return -EINVAL for default case ?

dbaluta replied:
Fixed.

RanderWang reviewed (commented):
LGTM for most part

dbaluta commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,331 @@
+	micfil_set_quality(dai);
+
+	/* set default gain to 2 */
+	dai_write(dai, REG_MICFIL_OUT_CTRL, 0x22222222);
```
fixed

dbaluta commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,331 @@
+			MICFIL_FIFO_CTRL_FIFOWMK_BITS(31));
+
+	/* enable channels */
+	dai_update_bits(dai, REG_MICFIL_CTRL1, 0xff, ((1 << micfil->params.pdm_ch) - 1));
```
Good point.

dbaluta commented:
@iuliana-prodan  @kv2019i addressed all your comments.

@lyakh  @lgirdwood  it looks like there are some problems with Intel build with zephyr.

Care to have a look?

```
 In file included from /zep_workspace/zephyr/include/zephyr/toolchain.h:50,
                 from /zep_workspace/zephyr/include/zephyr/sys/time_units.h:10,
                 from /zep_workspace/zephyr/include/zephyr/sys/util.h:640,
                 from /zep_workspace/sof/src/include/sof/audio/module_adapter/iadk/adsp_stddef.h:15,
                 from /zep_workspace/sof/src/audio/module_adapter/iadk/module_initial_settings_concrete.cpp:7:
/zep_workspace/zephyr/include/zephyr/toolchain/gcc.h:87: error: "ZRESTRICT" redefined [-Werror]
   87 | #define ZRESTRICT __restrict
      | 
In file included from /zep_workspace/zephyr/include/zephyr/sys/util.h:18:
/zep_workspace/zephyr/include/zephyr/toolchain/common.h:33: note: this is the location of the previous definition
   33 | #define ZRESTRICT
      | 
cc1plus: all warnings being treated as errors
[252/361] Building C object modules/sof/CMakeFiles/modules_sof.dir/zep_workspace/sof/src/audio/module_adapter/module/modules.c.obj
[253/361] Building C object modules/sof/CMakeFiles/modules_sof.dir/zep_workspace/sof/src/audio/module_adapter/module_adapter.c.obj
ninja: build stopped: subcommand failed.
```

dbaluta commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,335 @@
+		qsel = MICFIL_QSEL_VLOW2_QUALITY;
+		break;
+	default:
+		return -EINVAL;
```
Yes, both cases are returning -EINVAL. Do you see any problem here?

iuliana-prodan replied:
Add an error message to know what's wrong, the same as in _trigger function.

dbaluta replied:
Oh, I see. Agree. Fixed.

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+
+	switch (qsel) {
+	case MICFIL_QSEL_HIGH_QUALITY:
+		pdm_clk = rate * 8 * osr / 2; /* kfactor = 0.5 */
```
this is identical to the case below (8 / 2 = 4).

dbaluta replied:
Yes, this has the same final value! But, I prefer to keep it that way to keep this in line with the documentation.

lyakh replied:
Oh... This is very... strange. It's your area so it's up to you to decide, but... This looks very confusing to me. You have multiple `switch`-`case` entries for different quality values, you *seem* to have different formulae, ant comments specify different "kfactor" values, but the result is the same. At the **very least** I'd put a comment there, preferably I'd unify all the `case`s and provide just one formula, and just explain in the comment how that matches the documentation

dbaluta replied:
Yes, this comes from this table:

![image](https://github.com/thesofproject/sof/assets/62316/26b220d1-a8eb-41b9-b61e-09dbec16024f)

and this formula 

![image](https://github.com/thesofproject/sof/assets/62316/5581e6a1-c593-4e1f-9b6d-4f1727b8a4c0)


I will add more comments to better explain the code.

lyakh replied:
yeah, an explanation would be good, because even with that table I don't understand: e.g. with high quality the formula in the table seems to be
```
pdm_clk = rate * 8 * osr;
```
whereas you use
```
pdm_clk = rate * 8 * osr / 2;
```

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+		break;
+	case MICFIL_QSEL_LOW_QUALITY:
+	case MICFIL_QSEL_VLOW1_QUALITY:
+		pdm_clk = rate * 2 * osr * 2; /* kfactor = 2 */
```
and this is identical to the two above

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+		pdm_clk = rate * 2 * osr * 2; /* kfactor = 2 */
+		break;
+	case MICFIL_QSEL_VLOW2_QUALITY:
+		pdm_clk = rate * osr; /* kfactor = 4 */
```
can any of the above overflow to become negative?

dbaluta replied:
it shouldn't. We only support rates under 48000 and osr can be up to 256. I guess this function should return an unsigned value.

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+	int pdm_clk;
+
+	pdm_clk = micfil_get_pdm_clk(dai, rate);
+	if (pdm_clk <= 0)
```
`micfil_get_pdm_clk()` doesn't really return negative numbers - unless multiplication overflows, and it shouldn't

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+
+	clk_div = micfil_get_clk_div(dai, rate);
+	if (clk_div < 0)
+		return -EINVAL;
```
propagate the error

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+
+	ret = micfil_set_clock_params(dai, micfil->params.pdm_rate);
+	if (ret < 0)
+		return -EINVAL;
```
propagate

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+	dai_update_bits(dai, REG_MICFIL_STAT, 0xff, 0xff);
+
+	return 0;
+}
```
should this be `void`?

dbaluta replied:
yes, will fix.

lyakh commented on src/drivers/imx/micfil.c:
```diff
@@ -0,0 +1,336 @@
+	return 0;
+}
+
+static int micfil_stop(struct dai *dai)
```
`void`?

dbaluta replied:
Yes, will fix.

lyakh commented on src/ipc/ipc3/dai.c:
```diff
@@ -156,6 +160,8 @@ int ipc_dai_data_config(struct dai_data *dd, struct comp_dev *dev)
 		dd->stream_id = config->alh.stream_id;
 		break;
+	case SOF_DAI_IMX_MICFIL:
+		COMPILER_FALLTHROUGH;
```
usually we don't use `COMPILER_FALLTHROUGH` if there's no code between `case` lines - see lines 173-174 below

dbaluta replied:
I don't remember the reason we had these, but I just followed the existing code. I will remove it and have proper code.

dbaluta commented on zephyr/CMakeLists.txt:
```diff
@@ -239,6 +239,7 @@ if (CONFIG_SOC_SERIES_NXP_IMX8M)
 		${SOF_DRIVERS_PATH}/imx/sdma.c
 		${SOF_DRIVERS_PATH}/imx/sai.c
 		${SOF_DRIVERS_PATH}/imx/ipc.c
+		${SOF_DRIVERS_PATH}/imx/micfil.c
```
@lyakh thanks. i messed up when separated the patches :). will fix.

dbaluta commented:
@lyakh  all your comments should be addressed now

lyakh commented:
> @lyakh all your comments should be addressed now

What about https://github.com/thesofproject/sof/pull/8184#discussion_r1325413936 - you're sure the calculations in the code are correct? I cannot see how they match the table that you provided

kv2019i commented:
Feel free @dbaluta to merge

vijendarmukunda commented:
It seems kernel side changes not merged yet. We are going to upstream few patches on SOF side adding new DAI configuration for AMD platforms. Could you please push kernel side changes?

dbaluta commented:
@vijendarmukunda you are right, we are now in the middle of an internal release. Is it OK for you to delay this until next week?

vijendarmukunda commented:
@dbaluta : Sounds Okay. We will wait.

dbaluta commented:
@vijendarmukunda sent PR for review: https://github.com/thesofproject/linux/pull/4670
